File: frontend/src/types/RunLog.ts
// Interfaces for RunLog retrieval processes

/**
 * Interface for request to read FunctionModelVersion's RunLogs.
 */
export interface ReadVersionRunLogsRequest {
  function_model_version_uuid: string;
}

/**
 * Interface for request to read Project's RunLogs.
 */
export interface ReadProjectRunLogsRequest {
  project_uuid: string;
  page: number;
  rows_per_page: number;
}

/**
 * Interface for request to read BatchRun's RunLogs.
 */
export interface ReadBatchRunLogsRequest {
  batch_run_uuid: string;
  page: number;
  rows_per_page: number;
}

/**
 * Interface for request to read Project's RunLogs count.
 */
export interface ReadProjectRunLogsCountRequest {
  project_uuid: string;
}

/**
 * Interface for response of RunLogs count.
 */
export interface ReadProjectRunLogsCountResponse {
  project_uuid: string;
  count: number;
}

/**
 * General interface for RunLog.
 */
export interface RunLog {
  id: number;
  created_at: string;
  uuid: string;
  run_from_deployment: boolean;
  input_register_name: string;
  inputs: Record<string, any>;
  raw_output: string;
  parsed_outputs: Record<string, any>;
  function_call: Record<string, any>;
  run_log_metadata: Record<string, any>;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  latency: number;
  cost: number;
  version_uuid: string;
  project_uuid: string;
  sample_input_uuid?: string;
}

/**
 * Interface for RunLog with score.
 */
export interface RunLogWithScore extends RunLog {
  score: number;
  eval_metric_name: string;
  eval_metric_uuid: number;
}

export interface saveRunLogsRequest {
  inputs?: Record<string, any>;
  raw_output?: string;
  parsed_outputs?: Record<string, any>;
  tool_calls?: Array<Record<string, any>>;
  function_call?: Record<string, any>;

  sample_input_uuid?: string;
  run_log_metadata?: Record<string, any>;
}

File: backend/api/web/models/run_log.py
"""APIs for RunLog"""
from typing import Any, Dict, Optional, List

from modules.types import PMObject


class RunLogInstance(PMObject):
    id: int
    created_at: str
    uuid: str

    run_from_deployment: bool

    inputs: Optional[Dict[str, Any]] = None
    raw_output: Optional[str] = None
    parsed_outputs: Optional[Dict[str, Any]] = None
    tool_calls: Optional[List[Dict[str, Any]]] = None
    function_call: Optional[Dict[str, Any]] = None

    run_log_metadata: Optional[Dict[str, Any]] = None

    latency: Optional[float] = None
    cost: Optional[float] = None
    prompt_tokens: Optional[int] = None
    completion_tokens: Optional[int] = None
    total_tokens: Optional[int] = None

    version_uuid: str
    project_uuid: str
    sample_input_uuid: Optional[str] = None
    
class SaveRunLogBody(PMObject):
    inputs: Optional[Dict[str, Any]] = None
    raw_output: Optional[str] = None
    parsed_outputs: Optional[Dict[str, Any]] = None
    tool_calls: Optional[List[Dict[str, Any]]] = None
    function_call: Optional[Dict[str, Any]] = None
    
    sample_input_uuid: Optional[str] = None
    run_log_metadata: Optional[Dict[str, Any]] = None

class DeploymentRunLogViewInstance(PMObject):
    run_log_uuid: str
    project_uuid: str

    function_model_uuid: str
    function_model_name: str

    function_model_version_uuid: str
    function_model_version: int

    created_at: str

    inputs: Optional[Dict[str, Any]] = None
    raw_output: Optional[str] = None
    parsed_outputs: Optional[Dict[str, Any]] = None
    tool_calls: Optional[List[Dict[str, Any]]] = None
    function_call: Optional[Dict[str, Any]] = None

    prompt_tokens: Optional[int] = None
    completion_tokens: Optional[int] = None
    total_tokens: Optional[int] = None
    latency: Optional[float] = None
    cost: Optional[float] = None

    run_log_metadata: Optional[Dict[str, Any]] = None

    run_from_deployment: bool = True


class RunLogsCountInstance(PMObject):
    project_uuid: str
    count: int
